clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3783,6 +3612,363 @@ static void handleSYCLIntelMaxWorkGroupSize(Sema &S, Decl *D,
                                      AL.getArgAsExpr(1), AL.getArgAsExpr(2));
 }
 
+// Handles reqd_work_group_size.
+// If the reqd_work_group_size attribute is specified on
+// a declaration along with reqd_work_group_size attribute,
+// check to see if values of reqd_work_group_size arguments are
+// equal or greater than values of reqd_work_group_size attribute arguments.
+static bool checkReqdWorkGroupSizeAttrValues(
+    Sema &S, const Expr *RWGSXDim, const Expr *RWGSYDim, const Expr *RWGSZDim,
+    const Expr *R1WGSXDim, const Expr *R1WGSYDim, const Expr *R1WGSZDim) {
+  // If any of the operand is still value dependent, we can't test anything.
+  const auto *RWGSXDimExpr = dyn_cast<ConstantExpr>(RWGSXDim);
+  const auto *RWGSYDimExpr = dyn_cast<ConstantExpr>(RWGSYDim);
+  const auto *RWGSZDimExpr = dyn_cast<ConstantExpr>(RWGSZDim);
+  const auto *R1WGSXDimExpr = dyn_cast<ConstantExpr>(R1WGSXDim);
+  const auto *R1WGSYDimExpr = dyn_cast<ConstantExpr>(R1WGSYDim);
+  const auto *R1WGSZDimExpr = dyn_cast<ConstantExpr>(R1WGSZDim);
+
+  if (!RWGSXDimExpr || !RWGSYDimExpr || !RWGSZDimExpr || !R1WGSXDimExpr ||
+      !R1WGSYDimExpr || !R1WGSZDimExpr)
+    return false;
+
+  // Otherwise, check if value of reqd_work_group_size argument is
+  // less than value of reqd_work_group_size attribute argument.
+  bool CheckFirstArgument = S.getLangOpts().OpenCL
```

Comment:
Am I correct that the only difference between `checkWorkGroupSizeAttrValues()` and `checkReqdWorkGroupSizeAttrValues()` is that we use `>` in the first and `<` in the second when checking arguments? If so, we could use a template that takes a comparator as a template argument and get rid of one of these functions.

Reply:
>>Am I correct that the only difference between checkWorkGroupSizeAttrValues() and checkReqdWorkGroupSizeAttrValues() is that we use > in the first and < in the second when checking arguments?

Yes, correct

>>so, we could use a template that takes a comparator as a template argument and get rid of one of these functions.

Good suggestion. I will make the change. Thank you.

Reply:
Done